byzcoin/service_test.go

```diff
@@ -2194,102 +2193,74 @@ func TestService_StateChangeStorage(t *testing.T) {
 	}
 }
 
-// This tests that the service will restore the state changes
-// after a (re)boot and catch up potential new blocks
-func TestService_StateChangeCatchUp(t *testing.T) {
+// Tests that the state change storage will be caught up by a new conode
+func TestService_StateChangeStorageCatchUp(t *testing.T) {
+	// we don't want a db download
+	catchupDownloadAll = 100
+
 	s := newSer(t, 1, testInterval)
 	defer s.local.CloseAll()
 
-	n := 5
-	contract := func(cdb ReadOnlyStateTrie, inst Instruction, c []Coin) ([]StateChange, []Coin, error) {
-		// Check the state trie is created from the known global state
-		iid := inst.Hash()
-		if !bytes.Equal(inst.InstanceID.Slice(), s.darc.GetBaseID()) {
-			iid = inst.InstanceID.Slice()
-		}
-		_, ver, _, _, err := cdb.GetValues(iid)
-		sc1 := StateChange{
-			StateAction: Update,
-			InstanceID:  iid,
-			ContractID:  stateChangeCacheContract,
-			Version:     ver + 1,
-		}
-		if err != nil {
-			sc1.StateAction = Create
-		}
-		return []StateChange{sc1}, []Coin{}, nil
-	}
-	for _, s := range s.hosts {
-		RegisterContract(s, stateChangeCacheContract, adaptorNoVerify(contract))
-	}
-
-	createTx := func(iid []byte, counter uint64, wait int) *Instruction {
-		instr := Instruction{
-			InstanceID:       NewInstanceID(iid),
-			Spawn:            &Spawn{ContractID: stateChangeCacheContract},
-			SignerIdentities: []darc.Identity{s.signer.Identity()},
-			SignerCounter:    []uint64{counter},
-		}
-		tx := ClientTransaction{Instructions: Instructions{instr}}
-		err := tx.Instructions[0].SignWith(tx.Instructions.Hash(), s.signer)
+	n := 2
+	for i := 0; i < n; i++ {
+		tx, err := createClientTxWithTwoInstrWithCounter(s.darc.GetBaseID(), dummyContract, []byte{}, s.signer, uint64(i*2+1))
 		require.Nil(t, err)
 
 		_, err = s.service().AddTransaction(&AddTxRequest{
 			Version:       CurrentVersion,
 			SkipchainID:   s.genesis.SkipChainID(),
 			Transaction:   tx,
-			InclusionWait: wait,
+			InclusionWait: 10,
 		})
 		require.Nil(t, err)
-
-		return &instr
 	}
 
-	instr := createTx(s.darc.GetBaseID(), uint64(1), 1)
+	newServer, newRoster, newService := s.local.MakeSRS(cothority.Suite, 1, ByzCoinID)
+	registerDummy(newServer)
 
-	for i := 0; i < n-1; i++ {
-		// add transactions that must be recreated
-		createTx(instr.Hash(), uint64(i+1), 0)
-	}
-	createTx(instr.Hash(), uint64(n), 2)
+	newRoster = onet.NewRoster(append(s.roster.List, newRoster.List...))
+	ctx, _ := createConfigTxWithCounter(t, testInterval, *newRoster, defaultMaxBlockSize, s, 5)
+	s.sendTxAndWait(t, ctx, 10)
 
-	// Remove some entries to check it will recreate them
-	err := s.service().stateChangeStorage.db.Update(func(tx *bbolt.Tx) error {
-		b := s.service().stateChangeStorage.getBucket(tx, s.genesis.SkipChainID())
-		if b == nil {
-			return errors.New("missing bucket")
-		}
+	for i := 0; i < 10; i++ {
+		log.Lvl1("Sleeping for 1 second...")
+		time.Sleep(time.Second)
```

Comment:
is this because we're waiting for the nodes to catch up?

Reply:
Yes for the new one added